File: backend/detection/pipeline.py
"""
Main detection pipeline combining rule-based and ML scoring.
Orchestrates the full detection workflow.
"""

from typing import Optional
from backend.detection.rule_engine import get_rule_engine
from backend.detection.ml_scorer import get_ml_scorer
from backend.events.models import DetectionResult


class DetectionPipeline:
    """
    Combined detection pipeline using rules (60%) and ML (40%).
    """

    def __init__(self, rule_weight: float = 0.6, ml_weight: float = 0.4):
        """
        Initialize detection pipeline.
        
        Args:
            rule_weight: Weight for rule-based score (default 60%)
            ml_weight: Weight for ML score (default 40%)
        """
        if abs((rule_weight + ml_weight) - 1.0) > 0.001:
            raise ValueError(f"Weights must sum to 1.0, got {rule_weight + ml_weight}")
        
        self.rule_weight = rule_weight
        self.ml_weight = ml_weight
        self.rule_engine = get_rule_engine()
        self.ml_scorer = None
        
        # Try to load ML scorer, but don't fail if model not available yet
        try:
            self.ml_scorer = get_ml_scorer()
        except FileNotFoundError:
            print("⚠️  ML model not yet trained. Detection will use rules only.")
            
        self.suspicious_threshold = 25.0
        self.malicious_threshold = 60.0

    def update_thresholds(self, suspicious: float, malicious: float):
        self.suspicious_threshold = suspicious
        self.malicious_threshold = malicious

    def detect(self, command: str, process_memory_mb: float = 0.0, system_memory_percent: float = 0.0) -> DetectionResult:
        """
        Analyze a command and return detection result.
        
        Args:
            command: Command string to analyze
            process_memory_mb: Instantaneous memory allocation
            system_memory_percent: Total system RAM usage
            
        Returns:
            DetectionResult with risk score and classification
        """
        # Get rule-based score
        rule_score, matched_rules = self.rule_engine.score_rules(
            command, 
            process_memory_mb=process_memory_mb, 
            system_memory_percent=system_memory_percent
        )
        
        # Get ML score
        ml_score = 0.0
        ml_confidence = 0.0
        if self.ml_scorer:
            ml_score, ml_confidence = self.ml_scorer.score_ml(command)
        
        # Combine scores (weighted average)
        if self.ml_scorer:
            combined_score = (self.rule_weight * rule_score) + (self.ml_weight * ml_score)
        else:
            # If ML not available, use rules only
            combined_score = rule_score
        
        # Classify based on score
        if combined_score < self.suspicious_threshold:
            classification = "safe"
        elif combined_score < self.malicious_threshold:
            classification = "suspicious"
        else:
            classification = "malicious"
        
        # Build explanation
        explanation = self._build_explanation(
            classification, combined_score, matched_rules, ml_confidence
        )
        
        return DetectionResult(
            risk_score=combined_score,
            classification=classification,
            matched_rules=matched_rules,
            ml_confidence=ml_confidence,
            explanation=explanation,
        )

    def _build_explanation(
        self,
        classification: str,
        risk_score: float,
        matched_rules: list,
        ml_confidence: float,
    ) -> str:
        """
        Build a human-readable explanation of the detection.
        
        Args:
            classification: The classification result
            risk_score: The risk score (0-100)
            matched_rules: List of triggered rules
            ml_confidence: ML model confidence
            
        Returns:
            Explanation string
        """
        parts = []
        
        # Classification explanation
        if classification == "safe":
            parts.append("✅ Command appears safe.")
        elif classification == "suspicious":
            parts.append("⚠️  Command is suspicious and may pose a risk.")
        else:  # malicious
            parts.append("🚨 Command is likely malicious and should be blocked.")
        
        parts.append(f"Risk Score: {risk_score:.1f}/100")
        
        # Matched rules
        if matched_rules:
            parts.append(f"Detected patterns: {', '.join(matched_rules)}")
        else:
            parts.append("No suspicious patterns detected in command.")
        
        # ML confidence
        if self.ml_scorer:
            parts.append(f"ML Model confidence: {ml_confidence*100:.1f}%")
        
        return " | ".join(parts)


# Singleton instance
_detection_pipeline = None


def get_detection_pipeline(
    rule_weight: float = 0.6, ml_weight: float = 0.4
) -> DetectionPipeline:
    """
    Get or create the detection pipeline singleton.
    
    Args:
        rule_weight: Weight for rule-based score
        ml_weight: Weight for ML score
        
    Returns:
        The global DetectionPipeline instance
    """
    global _detection_pipeline
    if _detection_pipeline is None:
        _detection_pipeline = DetectionPipeline(rule_weight, ml_weight)
    return _detection_pipeline

File: backend/internal/detector/pipeline.go
// Package detector — Detection pipeline combining Rule Engine (60%) + ML (40%).
//
// Matches Python DetectionPipeline exactly:
//   - Combined score = 0.6 × rule_score + 0.4 × ml_score
//   - If ML is unavailable, combined score = rule_score (100% rules)
//   - Thresholds: suspicious ≥ 25, malicious ≥ 60
//   - Returns a *model.DetectionResult for each analysed command
package detector

import (
	"fmt"
	"log/slog"
	"strings"
	"sync"

	"github.com/Raphel6969/Kernal_AI_Security/backend/internal/model"
)

// Pipeline is the combined threat-detection pipeline.
// It is safe for concurrent use.
type Pipeline struct {
	rules      *RuleEngine
	ml         *MLScorer
	ruleWeight float64
	mlWeight   float64

	mu                  sync.RWMutex // protects the two threshold fields
	SuspiciousThreshold float64
	MaliciousThreshold  float64
}

var (
	pipelineOnce      sync.Once
	pipelineSingleton *Pipeline
)

// GetPipeline returns (or initialises) the package-level Pipeline singleton.
// modelPath points to data/trained_model.json; pass "" to disable ML scoring.
func GetPipeline(modelPath string) *Pipeline {
	pipelineOnce.Do(func() {
		pipelineSingleton = newPipeline(modelPath)
	})
	return pipelineSingleton
}

func newPipeline(modelPath string) *Pipeline {
	p := &Pipeline{
		rules:               GetRuleEngine(),
		ruleWeight:          0.6,
		mlWeight:            0.4,
		SuspiciousThreshold: 25.0,
		MaliciousThreshold:  60.0,
	}

	if modelPath != "" {
		p.ml = GetMLScorer(modelPath)
		if !p.ml.IsReady() {
			slog.Warn("Pipeline: ML scorer unavailable — rules-only mode")
		}
	} else {
		slog.Info("Pipeline: no model path provided — rules-only mode")
	}

	return p
}

// UpdateThresholds adjusts the classification thresholds at runtime.
// Thread-safe.
func (p *Pipeline) UpdateThresholds(suspicious, malicious float64) {
	p.mu.Lock()
	defer p.mu.Unlock()
	p.SuspiciousThreshold = suspicious
	p.MaliciousThreshold = malicious
	slog.Info("Pipeline: thresholds updated",
		"suspicious", suspicious, "malicious", malicious)
}

// GetThresholds returns the current thresholds.
func (p *Pipeline) GetThresholds() (suspicious, malicious float64) {
	p.mu.RLock()
	defer p.mu.RUnlock()
	return p.SuspiciousThreshold, p.MaliciousThreshold
}

// Detect analyses a command and returns a fully populated DetectionResult.
//
// Parameters:
//   - command     — the raw command string to analyse
//   - processMem  — instantaneous process RSS in MB (0 if unavailable)
//   - systemMem   — system-wide RAM usage percentage  (0 if unavailable)
func (p *Pipeline) Detect(command string, processMem, systemMem float64) *model.DetectionResult {
	// ── Tier A: Rule Engine ───────────────────────────────────────────────────
	ruleScore, matchedRules := p.rules.Score(command, processMem, systemMem)

	// ── Tier B: ML Scorer ─────────────────────────────────────────────────────
	var mlScore, mlConfidence float64
	mlReady := p.ml != nil && p.ml.IsReady()
	if mlReady {
		mlScore, mlConfidence = p.ml.Score(command)
	}

	// ── Weighted combination ──────────────────────────────────────────────────
	var combined float64
	if mlReady {
		combined = p.ruleWeight*ruleScore + p.mlWeight*mlScore
	} else {
		combined = ruleScore // rules-only fallback
	}
	if combined > 100 {
		combined = 100
	}

	// ── Classification ────────────────────────────────────────────────────────
	p.mu.RLock()
	suspT := p.SuspiciousThreshold
	malT := p.MaliciousThreshold
	p.mu.RUnlock()

	var classification string
	switch {
	case combined >= malT:
		classification = "malicious"
	case combined >= suspT:
		classification = "suspicious"
	default:
		classification = "safe"
	}

	// ── Explanation ───────────────────────────────────────────────────────────
	explanation := buildExplanation(classification, combined, matchedRules, mlConfidence, mlReady)

	return &model.DetectionResult{
		RiskScore:      combined,
		Classification: classification,
		MatchedRules:   matchedRules,
		MLConfidence:   mlConfidence,
		Explanation:    explanation,
	}
}

// ── Explanation builder ───────────────────────────────────────────────────────

// buildExplanation produces a human-readable detection summary.
// Output format is identical to the Python _build_explanation method so the
// React frontend displays the same strings.
func buildExplanation(
	classification string,
	riskScore float64,
	matchedRules []string,
	mlConfidence float64,
	mlReady bool,
) string {
	var parts []string

	switch classification {
	case "safe":
		parts = append(parts, "✅ Command appears safe.")
	case "suspicious":
		parts = append(parts, "⚠️  Command is suspicious and may pose a risk.")
	default: // malicious
		parts = append(parts, "🚨 Command is likely malicious and should be blocked.")
	}

	parts = append(parts, fmt.Sprintf("Risk Score: %.1f/100", riskScore))

	if len(matchedRules) > 0 {
		parts = append(parts, "Detected patterns: "+strings.Join(matchedRules, ", "))
	} else {
		parts = append(parts, "No suspicious patterns detected in command.")
	}

	if mlReady {
		parts = append(parts, fmt.Sprintf("ML Model confidence: %.1f%%", mlConfidence*100))
	}

	return strings.Join(parts, " | ")
}
